File: Blali(old stuff)/Framework/Tweening/Transition.vb


Imports Microsoft.Xna.Framework

Namespace Framework.Tweening

    <TestState(TestState.Finalized)>
    Public Class Transition(Of T)
        Implements ITransition

        Sub New(TransitionMethod As ITransitionType, StartValue As T, EndValue As T, FinishAction As FinishedDelegate)
            'Check whether the tweening manager supports this type
            Dim type As Type = GetType(T)
            If Automator.m_mapManagedTypes.ContainsKey(type) Then
                ManagedType = Automator.m_mapManagedTypes(type)

                Me.StartValue = StartValue
                Me.EndValue = EndValue
                Me.Method = TransitionMethod
                Me.FinishAction = FinishAction
                Me.TransitionStater = TransitionState.Idle
                Me.Value = StartValue
            Else
                Throw New NotImplementedException("The tweening manager doesn't support " & GetType(T).Name & ".")
            End If
        End Sub

        Sub New()
            'Check whether the tweening manager supports this type
            Dim type As Type = GetType(T)
            If Automator.m_mapManagedTypes.ContainsKey(type) Then
                ManagedType = Automator.m_mapManagedTypes(type)

                Me.TransitionStater = TransitionState.Idle
            Else
                Throw New NotImplementedException("The tweening manager doesn't support " & GetType(T).Name & ".")
            End If
        End Sub

        Sub Update(gameTime As GameTime) Implements ITransition.Update
            If Enabled And TransitionStater = TransitionState.InProgress Then
                Timer += gameTime.ElapsedGameTime.TotalMilliseconds

                'Calculate values
                Dim percentage As Double
                Dim completed As Boolean
                Method.onTimer(Timer, percentage, completed)

                'Set value
                Value = ManagedType.copy(ManagedType.getIntermediateValue(StartValue, EndValue, percentage))

                If completed Then
                    Select Case Repeat
                        Case RepeatJob.None
                            TransitionStater = TransitionState.Done
                        Case RepeatJob.Reverse
                            'Swap Start and Stop values
                            Dim tEnd As T = ManagedType.copy(EndValue)
                            EndValue = ManagedType.copy(StartValue)
                            StartValue = tEnd
                            'Reset Values
                            Value = ManagedType.copy(StartValue)
                            Timer = 0
                        Case RepeatJob.JumpBack
                            'Reset Values
                            Value = ManagedType.copy(StartValue)
                            Timer = 0
                    End Select
                    TriggerAction()
                End If
            End If
        End Sub

        Private Sub TriggerAction()
            If FinishAction IsNot Nothing Then FinishAction.Invoke(Me)
            RaiseEvent TransitionCompletedEvent(Me, New EventArgs)
        End Sub


        Public Property StartValue As T
        Public Property EndValue As T
        Public Property Value As T
        Public Property Method As ITransitionType Implements ITransition.Method
        Public Property FinishAction As FinishedDelegate 'A delegate to be executed when the transition is complete/the transition loops
        Public Property Enabled As Boolean = True
        Public Property Repeat As RepeatJob = RepeatJob.None
        Public Property TransitionStater As TransitionState Implements ITransition.TransitionStater
        Public ReadOnly Property ElapsedTime As Integer
            Get
                Return Timer
            End Get
        End Property

        Private Timer As Integer 'Keeps track of the elapsed time
        Private ManagedType As IManagedType 'Interface for converting/calculating values for the specified type

        Public Event TransitionCompletedEvent(sender As Object, e As EventArgs) 'An event to be executed when the transition is complete/the transition loops
        Public Delegate Sub FinishedDelegate(sender As Transition(Of T)) 'A delegate to be executed when the transition is complete/the transition loops

    End Class

End Namespace

File: Blali(old stuff)/Framework/Tweening/Transition.cs
public class Transition<T> : Tweening.ITransition
{
    public Transition(Tweening.ITransitionType TransitionMethod, T StartValue, T EndValue, FinishedDelegate FinishAction)
    {
        // Check whether the tweening manager supports this type
        var type = typeof(T);
        if (Emmond.Program.Automator.m_mapManagedTypes.ContainsKey(type))
        {
            ManagedType = Emmond.Program.Automator.m_mapManagedTypes[type];
            this.StartValue = StartValue;
            this.EndValue = EndValue;
            Method = TransitionMethod;
            this.FinishAction = FinishAction;
            TransitionStater = Tweening.TransitionState.Idle;
            Value = StartValue;
        }
        else
        {
            throw new NotImplementedException("The tweening manager doesn't support " + typeof(T).Name + ".");
        }
    }

    public Transition()
    {
        // Check whether the tweening manager supports this type
        var type = typeof(T);
        if (Emmond.Program.Automator.m_mapManagedTypes.ContainsKey(type))
        {
            ManagedType = Emmond.Program.Automator.m_mapManagedTypes[type];
            TransitionStater = Tweening.TransitionState.Idle;
        }
        else
        {
            throw new NotImplementedException("The tweening manager doesn't support " + typeof(T).Name + ".");
        }
    }

    public void Update(GameTime gameTime)
    {
        if (Enabled & TransitionStater == Tweening.TransitionState.InProgress)
        {
            Timer += gameTime.ElapsedGameTime.TotalMilliseconds;

            // Calculate values
            double percentage;
            bool completed;
            Method.onTimer(Timer, out percentage, out completed);

            // Set value
            Value = (T)ManagedType.copy(ManagedType.getIntermediateValue(StartValue, EndValue, percentage));
            if (completed)
            {
                var switchExpr = Repeat;
                switch (switchExpr)
                {
                    case global::Emmond.Framework.Tweening.RepeatJob.None:
                        {
                            StopTransition();
                            break;
                        }

                    case global::Emmond.Framework.Tweening.RepeatJob.Reverse:
                        {
                            // Swap Start and Stop values
                            T tEnd = (T)ManagedType.copy(EndValue);
                            EndValue = (T)ManagedType.copy(StartValue);
                            StartValue = tEnd;
                            // Reset Values
                            Value = (T)ManagedType.copy(StartValue);
                            Timer = 0;
                            break;
                        }

                    case global::Emmond.Framework.Tweening.RepeatJob.JumpBack:
                        {
                            // Reset Values
                            Value = (T)ManagedType.copy(StartValue);
                            Timer = 0;
                            break;
                        }
                }
            }
        }
    }

    private void StopTransition()
    {
        TransitionStater = Tweening.TransitionState.Done;
        if (FinishAction is object)
            FinishAction.Invoke(this);
        TransitionCompletedEvent?.Invoke(this, new EventArgs());
    }

    public T StartValue { get; set; }
    public T EndValue { get; set; }
    public T Value { get; set; }
    public Tweening.ITransitionType Method { get; set; }
    public FinishedDelegate FinishAction { get; set; } // A delegate to be executed when the transition is complete/the transition loops
    public bool Enabled { get; set; } = true;
    public Tweening.RepeatJob Repeat { get; set; } = Tweening.RepeatJob.None;
    public Tweening.TransitionState TransitionStater { get; set; }

    private int Timer; // Keeps track of the elapsed time
    private Tweening.IManagedType ManagedType; // Interface for converting/calculating values for the specified type

    public event TransitionCompletedEventEventHandler TransitionCompletedEvent;

    public delegate void TransitionCompletedEventEventHandler(object sender, EventArgs e); // An event to be executed when the transition is complete/the transition loops

    public delegate void FinishedDelegate(Transition<T> sender); // A delegate to be executed when the transition is complete/the transition loops
}
